scripts/vehicle_node.py

```diff
@@ -0,0 +1,380 @@
+        time = 0.05
+
+        #Calculate angle turned in the given time using omega = theta/time
+        angle = self.twist.angular.z*time
+
+        #Calculate distance travelled in the given time using linear velocity = arc distance/time
+        distance = self.twist.linear.x*time
+
+        #Calculate yaw, pitch and roll of the robot (pitch and roll currently not calculated)
+        self.pose.orientation.x = 0.0
+        self.pose.orientation.y = 0.0
+        self.pose.orientation.z += angle
+
+        #Calculate vehicle x, y, z position coordinates 
+        self.pose.position.x += (distance)*cos(self.pose.orientation.z)
+        self.pose.position.y += (distance)*sin(self.pose.orientation.z)
+
+        myarray = np.asarray(self.cloud)
+
+        points = []
+        values = []
+
+        for x in range(len(myarray)):
+            if (self.pose.position.x - 0.05 - resolution) <= myarray[x,0] <=  (self.pose.position.x + 0.05 + resolution):
+                if (self.pose.position.y - 0.05 - resolution) <= myarray[x,1] <= (self.pose.position.y + 0.05 + resolution):
+                    innerlist = []
+                    innerlist.append(myarray[x,0])
+                    innerlist.append(myarray[x,1])
+                    values.append(myarray[x,2])
+                    points.append(innerlist)
+
+        x1, y1 = np.meshgrid(np.linspace((self.pose.position.x-0.05), (self.pose.position.x+0.05), 50), np.linspace((self.pose.position.y-0.05), (self.pose.position.y+0.05), 50))
+                
+        grid_z2 = griddata(points, values, (x1, y1), method='linear')
+
+        flat_list = []
+
+        for sublist in grid_z2:
+            for item in sublist:
+                flat_list.append(item)
+                
+        avg = sum(flat_list)/len(flat_list)
+
+        self.pose.position.z = avg
+
+
+    def publish_pose_msg(self, event):
+        
+        #Header
+        self.pose_stamped.header.stamp = rospy.Time.now()
+
+        #Pose information
+        self.pose_stamped.pose.position.x = self.pose.position.x
+        self.pose_stamped.pose.position.y = self.pose.position.y
+        self.pose_stamped.pose.position.z = self.pose.position.z
+
+        #Convert Euler to Quarternion
+        q = tf.transformations.quaternion_from_euler(self.pose.orientation.x, self.pose.orientation.y, self.pose.orientation.z, 'sxyz')
+        
+        self.pose_stamped.pose.orientation.x = q[0]
+        self.pose_stamped.pose.orientation.y = q[1]
+        self.pose_stamped.pose.orientation.z = q[2]
+        self.pose_stamped.pose.orientation.w = q[3]
+
+        #Broadcast vehicle frame which is a child of the world frame
+        br = tf.TransformBroadcaster()
+        br.sendTransform((self.pose.position.x, self.pose.position.y, self.pose.position.z ), q, rospy.Time.now(),"vehicle_frame", "map")
```

Comment:
We have timing info from the event object, it's better to use that instead of rospy.Time.now()